src/WriteClient.ts

```diff
@@ -0,0 +1,906 @@
+				try {
+					url = new URL(file)
+				} catch (error) {
+					// noop only on invalid URL, fetch errors will throw in the next if statement
+				}
+
+				if (url) {
+					// File is a URL, fetch it
+					resolvedFile = await this.fetchForeignAsset(
+						url.toString(),
+						fetchParams,
+					)
+				} else {
+					// File is actual file content, use it as-is
+					resolvedFile = file
+				}
+			} else if (file instanceof URL) {
+				// File is a URL instance, fetch it
+				resolvedFile = await this.fetchForeignAsset(
+					file.toString(),
+					fetchParams,
+				)
+			} else {
+				resolvedFile = file
+			}
+
+			const asset = await this.createAsset(resolvedFile, filename, {
+				...{ notes, credits, alt, tags },
+				...fetchParams,
+			})
+
+			migrationAsset.asset = asset
+		}
+
+		reporter?.({
+			type: "assets:created",
+			data: {
+				created,
+			},
+		})
+	}
+
+	/**
+	 * Creates documents in the Prismic repository's migration release.
+	 *
+	 * @param migration - A migration prepared with {@link createMigration}.
+	 * @param params - An event listener and additional fetch parameters.
+	 *
+	 * @internal This method is one of the step performed by the {@link migrate} method.
+	 */
+	private async migrateCreateDocuments(
+		migration: Migration<TDocuments>,
+		{
+			reporter,
+			...fetchParams
+		}: { reporter?: (event: MigrateReporterEvents) => void } & FetchParams = {},
+	): Promise<void> {
+		// Resolve master locale
+		const repository = await this.getRepository(fetchParams)
+		const masterLocale = repository.languages.find((lang) => lang.is_master)!.id
+		reporter?.({
+			type: "documents:masterLocale",
+			data: {
+				masterLocale,
+			},
+		})
+
+		const documentsToCreate: PrismicMigrationDocument<TDocuments>[] = []
+		// We create an array with non-master locale documents last because
+		// we need their master locale document to be created first.
+		for (const doc of migration._documents) {
+			if (!doc.document.id) {
+				if (doc.document.lang === masterLocale) {
+					documentsToCreate.unshift(doc)
+				} else {
+					documentsToCreate.push(doc)
+				}
+			}
+		}
+
+		let created = 0
+		for (const doc of documentsToCreate) {
+			reporter?.({
+				type: "documents:creating",
+				data: {
+					current: ++created,
+					remaining: documentsToCreate.length - created,
+					total: documentsToCreate.length,
+					document: doc,
+				},
+			})
+
+			// Resolve master language document ID for non-master locale documents
+			let masterLanguageDocumentID: string | undefined
+			if (doc.masterLanguageDocument) {
+				const masterLanguageDocument =
+					await resolveMigrationContentRelationship(doc.masterLanguageDocument)
+
+				masterLanguageDocumentID =
+					"id" in masterLanguageDocument ? masterLanguageDocument.id : undefined
+			} else if (doc.originalPrismicDocument) {
+				masterLanguageDocumentID =
+					doc.originalPrismicDocument.alternate_languages.find(
+						({ lang }) => lang === masterLocale,
+					)?.id
```

Comment:
❓ #ask: Do we need to handle the case where the found document doesn't exist? It looks like the ID given here will be the original Prismic ID, which will not exist unless the user explicitly updated it (unlikely).

Reply:
Ohhh, nice catch, this is definitely an issue!

Reply:
Fixed!